api/coreservice.go

```diff
@@ -1379,20 +1386,51 @@ func (core *coreService) LogsInRange(filter *logfilter.LogFilter, start, end, pa
 		return nil, err
 	}
 
-	// TODO: improve using goroutine
 	if paginationSize == 0 {
 		paginationSize = 1000
 	}
 	if paginationSize > 5000 {
 		paginationSize = 5000
 	}
+
+	c := make(chan _logsInBlock)
 	logs := []*iotextypes.Log{}
+	if len(blockNumbers) == 0 {
+		return logs, nil
+	}
 	for _, i := range blockNumbers {
-		logsInBlock, err := core.LogsInBlock(filter, i)
-		if err != nil {
+		go func(blockNumber uint64) {
+			logsInBlock, err := core.logsInBlock(filter, blockNumber)
+			ret := &_logsInBlock{
+				blockNumber,
+				logsInBlock,
+				err,
+			}
+			c <- *ret
+		}(i)
+	}
+
+	i := 0
+	orderedBlocks := []_logsInBlock{}
+	for logsInBlock := range c {
+		if logsInBlock.err != nil {
 			return nil, err
 		}
-		for _, log := range logsInBlock {
+
+		orderedBlocks = append(orderedBlocks, logsInBlock)
+
+		if i == len(blockNumbers)-1 {
+			close(c)
+		}
+		i++
+	}
+
+	sort.SliceStable(orderedBlocks, func(i, j int) bool {
```

Comment:
Sorting is not encouraged. IMO, order can be achieved by `producer-consumer` model or by assigning the ans array `logs[i]` in the gorouine `i`, or a new design by yourself

Reply:
The producer-consumer model should prevent `cpu` from consuming out of resources.